R/pkg-arrow.R

```diff
@@ -2,24 +2,41 @@
 ArrowSubstraitCompiler <- R6::R6Class(
   "ArrowSubstraitCompiler",
   inherit = SubstraitCompiler,
+  private = list(extension_uri = NULL),
   public = list(
     initialize = function(...) {
+      private$extension_uri <- list(
+        "arithmetic" = substrait$extensions$SimpleExtensionURI$create(
+          extension_uri_anchor = 1L,
+          uri = "https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic.yaml"
+        ),
+        "comparison" = substrait$extensions$SimpleExtensionURI$create(
+          extension_uri_anchor = 2L,
+          uri = "https://github.com/substrait-io/substrait/blob/main/extensions/functions_comparison.yaml"
+        )
+      )
       super$initialize(...)
       self$.fns = arrow_funs
     },
+
+    extension_uri_anchor = function(name){
+      prefix = strsplit(name, ".", fixed = TRUE)[[1]][1]
+      private$extension_uri[[prefix]]$extension_uri_anchor
+    },
+
     evaluate = function(...) {
       plan <- self$plan()
 
-      # Here we only implement 'add()', so this works because the only
-      # function that we ever use is contained in this extensions definition.
-      plan$extension_uris[[1]]$uri <-
-        "https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic.yaml"
-
       substrait_eval_arrow(
         plan = plan,
         tables = self$named_table_list(),
         col_names = self$schema$names
       )
+    },
+
+    resolve_function  = function(name, args, template, output_type = NULL){
+      fun_name <- strsplit(name, ".", fixed = TRUE)[[1]][2]
+      super$resolve_function(fun_name, args, template, output_type)
```

Comment:
I made an update to switch the names for the consumer-friendly ones later now, but I looked at the plan and I've still only going a single extension URI anchor even when using functions from multiple ones. I'm wondering if the bit of Arrow code that consumes the plan even uses that field?

Reply:
I can change the URI value to any old nonsense and the code still runs successfully.

Reply:
Sorry, instead of "still runs" I should have said "still runs, and also returns the correct result".  I'm a bit stuck as I can't write a test as I don't know what "correctly" looks like.  And, for the sake of argument, if the results returned are correct regardless of the value in the field, wouldn't that mean that it's not been implemented properly in the consumer?

Reply:
Just in case future me (or someone else) comes back to this and wonders what happened here in the end, I had a private conversation with Dewey in which we realised:

- the code was previously modifying the compiler but needed to instead modify the plan
- the reason that we were getting NAs for the function names is that we were using strsplit to reassign values multiple times on the same values, and so the first time we ran `plan()` (before we fixed it) we'd get the correct function name, but the subsequent times we'd get NAs
- the plan shown by `compiler$plan()` isn't the same as the plan produced by `substrait_eval_arrow()` and so using the former as diagnostic of changes isn't going to help